[sram/dv] Update DV doc (#9914)
lowRISC/opentitan: hw/ip/sram_ctrl/doc/dv/index.md

It can be created manually by invoking [`regtool`]({{< relref "util/reggen/README.md" >}}):

Besides default RAL that provides the access to CSRs, there is another RAL for SRAM memory.

> This RAL is in `sram_ctrl_prim_ral_pkg.sv` which is added manually and enabled with all the common CSR/memory tests, such mem_walk, tl_errors, tl_intg_err etc.

Reviewer: ```suggestion
This RAL is in `sram_ctrl_prim_ral_pkg.sv`, it is added manually and enabled with all the common CSR and memory tests, such as mem_walk, tl_errors, tl_intg_err, etc.
```

Author: Updated, thanks

### Reference models
There are 3 DV reference models used in this testbench.

> - The SRAM_CTRL testbench uses a SystemVerilog golden model of its address and data scrambling to ensure that internal encryption and decryption are performed correctly.

Reviewer 2: ```suggestion
- The SRAM_CTRL testbench uses a SystemVerilog golden model for address and data scrambling to ensure that internal encryption and decryption perform correctly.
```
Do you think this is better? Just a suggestion :)

Author: Updated, thanks

This golden model, [`sram_scrambler_pkg`]({{<relref "hw/dv/sv/mem_bkdr_util/README.md" >}}) is tightly integrated with the mem_bkdr_util for ease of use such that we can choose to enable encryption on any backdoor access.
- The frontdoor check [`mem_model`]({{<relref "hw/dv/sv/mem_model/README.md" >}}) is also enabled in cip_base_scoreboard.
The model is invoked at the end of the transaction as follows:
  - If it’s write, store the data on the address in the memory array.
  - If it’s read and the address has been written, compare read value with stored value.
  - If it’s read and the address has NOT been written, skip checking.

There are some limitations in the frontdoor check as follows:

>   1. Can’t check ECC if encoding and decoding match to each other.

Reviewer 2: ```suggestion
  1. Can’t check ECC if encoding and decoding match.
```

Author: Updated. Thanks

  2. Can’t check the read value if the address hasn't been written after init or after key request.
  3. Can’t check the write value if the address isn’t read after the write.
  4. Not aware of any B2B hazard (such as RAW).

- The backdoor check [`mem_bkdr_scb`]({{<relref "hw/dv/sv/mem_bkdr_scb/README.md" >}}) is used to ensure frontdoor access matches with the backdoor expected value.
The [`sram_scrambler_pkg`]({{<relref "hw/dv/sv/mem_bkdr_util/README.md" >}}) is invoked in [`mem_bkdr_scb`]({{<relref "hw/dv/sv/mem_bkdr_scb/README.md" >}})
This backdoor check covers all the limitations of frontdoor check.
There is a limitation as follows, which can be covered by frontdoor check.

>   - Can’t check when memory is modified unintentionally after the access is completed.

Reviewer 2: I think this can still be checked by backdoor - say if we do a random backdoor read?
I would prefer to remove these two lines...

Author: backdoor check can't check if the data is preserved correctly. 
For example, if we write address 0 to 'hF and we read the backdoor value to compare, in order to make sure it's 'hF. 
Assume there is a design but, that value is modified to 'hE a few cycles after the write completes. We do a read and compare read data ('hE) with backdoor value. And the backdoor value will be fetched from memory, which is 'hE too. This change from 'hF -> 'hE will be caught by the front door check. So it will be fine.

#### Functional coverage
To ensure high quality constrained random stimulus, it is necessary to develop a functional coverage model.
The following covergroups have been developed to prove that the test intent has been adequately met:

> * b2b_access_types_cg (imported from mem_bkdr_scb): Cover that all combinations of reads/writes can create b2b scenario

Reviewer: Sorry, this description is not clear to me, do you mean back to back read and write requests to the same addresses? Also (nit), please add a closing period here and below.

Author: Updated, hope it is clear. Thanks

* sram_tl_d_chan_fifo: TL data channel for memory accesses
* kdi_fifo: For key refresh operations from OTP_CTRL

All CSR accesses made to the SRAM_CTRL register file are tracked and predicted by the scoreboard.

Verifying memory accesses is somewhat involved, and makes heavy use of the [`mem_bkdr_util`]({{< relref "hw/dv/sv/mem_bkdr_util/README.md" >}}).

At a high level, whenever a TL access is made to memory, all relevant information is stored in an `sram_trans_t` struct.

> Refer to section "Reference models" for the SRAM memory checking.

Reviewer: Some details about the mechanics of checking in the old text are omitted in the new text. Is that intentional?

Author: yes, old checking has been removed and the details are in `Reference models` section
